driver-core/src/main/com/mongodb/client/model/Filters.java

```diff
@@ -1040,14 +1040,31 @@ private void addClause(final BsonDocument document, final Map.Entry<String, Bson
         }
 
         private boolean keysIntersect(final BsonDocument first, final BsonDocument second) {
-            for (String name : first.keySet()) {
-                if (second.containsKey(name)) {
+            Set<String> firstKeySet = removeNegationFromKeySet(first);
+            Set<String> secondKeySet = removeNegationFromKeySet(second);
+            for (String name : firstKeySet) {
+                if (secondKeySet.contains(name)) {
                     return true;
                 }
             }
             return false;
         }
 
+        private Set<String> removeNegationFromKeySet(final BsonDocument document) {
+            if (document.containsKey("$not")) {
+                Set<String> keySet = new HashSet<String>();
```

Comment:
Do we need the other toplevel fields from the document if it contains $not ?

Reply:
This code has been removed.